source/extensions/filters/common/ext_authz/ext_authz_http_impl.h

```diff
@@ -12,7 +14,119 @@ namespace Filters {
 namespace Common {
 namespace ExtAuthz {
 
-typedef std::vector<std::pair<const Http::LowerCaseString, const std::string>> HeaderKeyValueVector;
+class Matcher;
+typedef std::shared_ptr<Matcher> MatcherSharedPtr;
+
+/**
+ *  Matchers describe the rules for matching authorization request and response headers.
+ */
+class Matcher {
+public:
+  virtual ~Matcher() {}
+
+  /**
+   * Returns whether or not the header key matches the rules of the matcher.
+   *
+   * @param key supplies the header key to be evaluated.
+   */
+  virtual bool matches(const Http::LowerCaseString& key) const PURE;
```

Comment:
Sorry why do we have these two different match functions taking different params?

Reply:
In some cases a `HeaderString` needs to be evaluated and `LowerCaseString` in others. I could use `std::string` as a common denominator, but then would be no guarantee that the input is a lower case string. That's what I was thinking.

Reply:
AFAICT they are always keys, right? That should always be lower case. Can you perhaps just use a string_view that we assume is lower case?

Reply:
Yes, in the case of this filter we are safe to assume that. Kind of unrelated, but I think we should add to Ext-Authz docs that the values used in these string matcher lists should be always lowercase.